Storage/src/Connection/Rest.php

```diff
@@ -628,6 +629,148 @@ private function sendWithRetry($resource, $method, $args)
+            $this->updateRetryInovationHeaders(
+                $arguments,
+                $requestHash
+            );
+        };
+
         return $this->send($resource, $method, $args);
     }
+
+    /**
+     * Updates the x-goog-api-client header value with UUID
+     * and retry count
+     *
+     * @return void
+     */
+    private function updateRetryInovationHeaders(
+        &$arguments,
+        $requestHash,
+        $currentAttempt = 0
+    ) {
+        $valueToAdd = sprintf("gccl-invocation-id/%s", $requestHash);
+        $this->updateHeader(
+            'x-goog-api-client',
+            $arguments,
+            $valueToAdd
+        );
+
+        $valueToAdd = sprintf("gccl-attempt-count/%s", $currentAttempt);
+        $this->updateHeader(
+            'x-goog-api-client',
+            $arguments,
+            $valueToAdd,
+            false
+        );
+    }
+
+    /**
+     * Amends the given header key with new value for a request such that
+     * the $request headers aren't modified directly and instead $options array
+     * which are applied to the request just before sending it at core level.
+     * Thus the $request object remains the same between each retry request at
+     * RequestWrappers' level.
+     *
+     * @param string $headerLine The header line to update.
+     * @param array &$arguments The arguments array(passed by reference) used by
+     * execute method of ExponentialBackoff object.
+     * @param string $value The value to be ammended in the header line.
+     * @param bool $getHeaderFromRequest [optional] A flag which determines if
+     *  existing header value is read from $request or from $options. It's useful
+     *  to read from $options incase we update multiple values to a single
+     *  header key.
+     */
+    private function updateHeader(
+        string $headerLine,
+        array &$arguments,
+        string $value,
+        bool $getHeaderFromRequest = true
+    ): void {
+        // Fetch request and options
+        $request = $this->fetchRequest($arguments);
+        $options = $this->fetchOptions($arguments);
+
+        // Create the modified header
+        $headerValue = '';
+        if ($getHeaderFromRequest) {
+            $headerValues = $request->getHeader($headerLine);
+            $headerValues[] = $value;
+            $headerValue = implode(' ', $headerValues);
+        } else {
+            $headerValue = $options['headers'][$headerLine];
+            $headerValue .= (' ' . $value);
+        }
+
+        // Amend the $option's header value
+        $options['headers'][$headerLine] = $headerValue;
+
+        // Set the $argument's options array
+        $this->setOptions($arguments, $options);
+    }
+
+
+    /**
+     * This helper method fetches Request object from the $argument list.
+     * @param mixed $arguments
+     * @return Request|null
+     */
+    private function fetchRequest($arguments): Request|null
```

Comment:
Can't use `Request|null` before 8.1

Reply:
Oh, got to know a new thing today😄. 

Thanks for catching and pointing this! Fixed in [33ba71f](https://github.com/googleapis/google-cloud-php/pull/5741/commits/33ba71f14ba1fd93fbfff19cb498b8e566f0d540)